File: python/0017.py
# Link: https://leetcode.com/problems/letter-combinations-of-a-phone-number/description/
# Runtime: 40 ms
# Memory: 16.42 MB

class Solution:
    buttons: Dict[str, str] = {}
    combinations: List[str] = []

    def __init__(self):
        self.buttons['2'] = 'abc'
        self.buttons['3'] = 'def'
        self.buttons['4'] = 'ghi'
        self.buttons['5'] = 'jkl'
        self.buttons['6'] = 'mno'
        self.buttons['7'] = 'pqrs'
        self.buttons['8'] = 'tuv'
        self.buttons['9'] = 'wxyz'

    def letterCombinations(self, digits: str) -> List[str]:
        if len(digits) == 0: return []

        self.combinations = []
        self.__add_combinations(digits, "", 0)

        return self.combinations

        
    def __add_combinations(self, digits: str, combination: str, depth: int) -> None:
        if len(digits) == depth: self.combinations.append(combination)
        else:
            for symbol in self.buttons[digits[depth]]:
                combination += symbol
                self.__add_combinations(digits, combination, depth + 1)
                combination = combination[:-1]

File: c++/0017.cpp
// Link: https://leetcode.com/problems/letter-combinations-of-a-phone-number/description/
// Runtime: 0 ms
// Memory: 6.5 MB

class Solution 
{
public:
    Solution()
    {
        buttons['2'] = "abc";
        buttons['3'] = "def";
        buttons['4'] = "ghi";
        buttons['5'] = "jkl";
        buttons['6'] = "mno";
        buttons['7'] = "pqrs";
        buttons['8'] = "tuv";
        buttons['9'] = "wxyz";
    }

    std::vector<std::string> letterCombinations(const std::string& digits) 
    {
        if (digits.empty()) return {};

        std::string combination = "";
        addCombinations(digits, combination, 0);

        return combinations;
    }

private:
    std::unordered_map<char, std::string> buttons;
    std::vector<std::string> combinations;

    void addCombinations(const std::string& digits, std::string& combination, int depth)
    {
        if (depth == static_cast<int>(digits.size())) combinations.push_back(combination);
        else
            for (const auto& symbol : buttons[digits[depth]])
            {
                combination.push_back(symbol);
                addCombinations(digits, combination, depth + 1);
                combination.pop_back();
            }
    }
};
